File: cases/taint/kotlin/context-pair-depth2-negative/ContextPairDepth2Negative.kt
package dataflowbench

object ContextPairDepth2Negative {
    fun dfb_source(): String { // DFB-SOURCE: context-pair-depth2-input
        return "tainted"
    }

    fun dfb_sink(value: String) {} // DFB-SINK: context-pair-depth2-sink

    fun helper(value: String): String { // DFB-WITNESS: context-pair-depth2-helper
        return value
    }

    fun wrapper(value: String): String { // DFB-WITNESS: context-pair-depth2-wrapper
        return helper(value)
    }

    fun outerTainted(): String {
        return wrapper(dfb_source())
    }

    fun outerClean(): String {
        return wrapper("clean")
    }

    fun run() {
        val tainted = outerTainted()
        val clean = outerClean()
        dfb_sink(clean)
    }
}

File: cases/taint/typescript/context-pair-depth2-negative/ContextPairDepth2Negative.ts
function dfb_source(): string { // DFB-SOURCE: context-pair-depth2-input
  return "tainted";
}

function dfb_sink(value: string): void {} // DFB-SINK: context-pair-depth2-sink

function helper(value: string): string { // DFB-WITNESS: context-pair-depth2-helper
  return value;
}

function wrapper(value: string): string {
  return helper(value);
}

function outerTainted(): string {
  return wrapper(dfb_source());
}

function outerClean(): string {
  return wrapper("clean");
}

function run(): void {
  const tainted: string = outerTainted();
  const clean: string = outerClean();
  dfb_sink(clean);
}
